front/src/common/Map/WarpedMap/core/helpers.ts

```diff
@@ -0,0 +1,173 @@
+export type GridIndex = Record<string, Triangle>;
+export type Grids = {
+  regular: GridFeature;
+  warped: GridFeature;
+};
+export type PointsGrid = Record<number, Position>[];
+
+/**
+ * Path manipulation helpers:
+ */
+export function getSamples(
+  line: Feature<LineString>,
+  samples: number
+): { step: number; points: Feature<Point>[] } {
+  if (samples <= 1) throw new Error('samples must be an integer greater than 1');
+
+  const points: Feature<Point>[] = [];
+  const l = length(line, { units: 'meters' });
+  const step = l / (samples - 1);
+  for (let i = 0; i < samples; i++) {
+    if (!i) {
+      points.push(point(first(line.geometry.coordinates) as Position));
+    } else if (i === samples - 1) {
+      points.push(point(last(line.geometry.coordinates) as Position));
+    } else {
+      const at = clamp(step * i, 0, l);
+      points.push(along(line, at, { units: 'meters' }));
+    }
+  }
+
+  return { step, points };
+}
+
+export function extendLine(line: Feature<LineString>, lengthToAdd: number): Feature<LineString> {
+  if (lengthToAdd <= 1) throw new Error('lengthToAdd must be a positive');
+
+  const points = line.geometry.coordinates;
+  const firstPoint = points[0] as Vec2;
+  const second = points[1] as Vec2;
+  const lastPoint = points[points.length - 1] as Vec2;
+  const secondToLast = points[points.length - 2] as Vec2;
+
+  return {
+    ...line,
+    geometry: {
+      ...line.geometry,
+      coordinates: [
+        vec.add(
+          firstPoint,
+          vec.multiply(
+            vec.vector(second, firstPoint),
+            lengthToAdd / distance(second, firstPoint, { units: 'meters' })
+          )
+        ),
+        ...points,
+        vec.add(
+          lastPoint,
+          vec.multiply(
+            vec.vector(secondToLast, lastPoint),
+            lengthToAdd / distance(secondToLast, lastPoint, { units: 'meters' })
+          )
+        ),
+      ],
+    },
+  };
+}
+
+/**
+ * Triangle helpers:
+ */
+export function getBarycentricCoordinates(
+  [x, y]: Position,
+  {
+    geometry: {
+      coordinates: [[[x0, y0], [x1, y1], [x2, y2]]],
+    },
+  }: Triangle
+): BarycentricCoordinates {
+  const denominator = (y1 - y2) * (x0 - x2) + (x2 - x1) * (y0 - y2);
+  const a = ((y1 - y2) * (x - x2) + (x2 - x1) * (y - y2)) / denominator;
+  const b = ((y2 - y0) * (x - x2) + (x0 - x2) * (y - y2)) / denominator;
+  const c = 1 - a - b;
+
+  return [a, b, c];
+}
+
+export function isInTriangle([a, b, c]: BarycentricCoordinates): boolean {
+  return a >= 0 && a <= 1 && b >= 0 && b <= 1 && c >= 0 && c <= 1;
+}
+
+export function getPointInTriangle(
+  [a, b, c]: BarycentricCoordinates,
+  {
+    geometry: {
+      coordinates: [[[x0, y0], [x1, y1], [x2, y2]]],
+    },
+  }: Triangle
+): Position {
+  return [a * x0 + b * x1 + c * x2, a * y0 + b * y1 + c * y2];
+}
+
+/**
+ * Data helpers:
+ */
+const OSRD_BATCH_SIZE = 500;
+export async function getImprovedOSRDData(
```

Comment:
Hahaha, I completely forgot it, sorry ^^
The explanations are here:

https://github.com/osrd-project/osrd/blob/2e4ed9e8172bf644462ea3062aff37dbd9016145/front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx#L239-L243

Basically, before we call that function, OSRD data we have were loaded from a `DataLoader` map. This means that they probably are simplified, because they come from "vector servers" that simplify GeoJSONs before sending them, so that the payloads are way smaller.

That's why when you open a map first you can get so many objects showing fast, but when you zoom quickly, they are loaded again, with better precision.

But once I know what are the OSRD entities to show, I can load the "exact value" of each of them, so that when the user zooms deeply in the final map, they see the best possible version of these entities.

Reply:
Ok, maybe we should add a comment to explain what you wrote here, but I understand better now :)